internal/sink/sink.go

```diff
@@ -55,3 +68,188 @@ func New(cfg config.Config) (Sink, error) {
+}
+
+func (f *fw) Close() error {
+	if f.w != nil {
+		if err := f.w.Close(); err != nil {
+			return err
+		}
+	}
+
+	if err := f.File.Close(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func NewFileWrapper(f *os.File, fmt config.Config, cmp config.Config) (*fw, error) {
+	var newline []byte
+	switch runtime.GOOS {
+	case "windows":
+		newline = []byte("\r\n")
+	default:
+		newline = []byte("\n")
+	}
+
+	// if the file format is not text-based, then newline is unused.
+	// if a file format uses a specific compression, then it should
+	// be configured and returned in this switch.
+	switch fmt.Type {
+	case "data":
+		newline = nil
+	}
+
+	switch cmp.Type {
+	case "gzip":
+		return &fw{f, newline, gzip.NewWriter(f)}, nil
+	case "snappy":
+		return &fw{f, newline, snappy.NewBufferedWriter(f)}, nil
+	case "zstd":
+		// TODO: add settings support
+		z, err := zstd.NewWriter(f)
+		if err != nil {
+			return nil, err
+		}
+
+		return &fw{f, newline, z}, nil
+	default:
+		return &fw{f, newline, nil}, nil
+	}
+}
+
+type filePath struct {
+	// Prefix is a prefix prepended to the file path.
+	//
+	// This is optional and has no default.
+	Prefix string `json:"prefix"`
+	// PrefixKey retrieves a value from an object that is used as
+	// the prefix prepended to the file path. If used, then
+	// this overrides Prefix.
+	//
+	// This is optional and has no default.
+	PrefixKey string `json:"prefix_key"`
+	// Suffix is a suffix appended to the file path and is used as
+	// the object filename.
+	//
+	// This is optional and has no default.
+	Suffix string `json:"suffix"`
+	// SuffixKey retrieves a value from an object that is used as
+	// the suffix appended to the file path. If used, then
+	// this overrides Suffix.
+	//
+	// This is optional and has no default.
+	SuffixKey string `json:"suffix_key"`
+	// TimeFormat inserts a formatted datetime string into the file path.
+	// The string uses pattern-based layouts
+	// (https://gobyexample.com/procTime-formatting-parsing).
+	//
+	// This is optional and has no default.
+	TimeFormat string `json:"time_format"`
+	// UUID inserts a random UUID into the file path. If a suffix is
+	// not set, then this is used as the filename.
+	//
+	// This is optional and defaults to false.
+	UUID bool `json:"uuid"`
+	// Extension appends a file extension to the filename.
+	//
+	// This is optional and defaults to false.
+	Extension bool `json:"extension"`
+}
+
+// New constructs a file path using the pattern
+// [prefix]/[time_format]/[uuid]/[suffix], where each field is optional
+// and builds on the previous field. The caller is responsible for
+// creating an OS agnostic file path (filepath.FromSlash is recommended).
+//
+// If only one field is set, then this constructs a filename,
+// otherwise it constructs a file path.
+//
+// If the struct is empty, then this returns an empty string. The caller is
+// responsible for creating a default file path if needed.
+func (p filePath) New() string {
```

Comment:
Test: could we add a test case for this logic that shows, verifies it can match selective data processing patterns/requirements? (like the previous default, AWS Glue, etc)

Reply:
I'll add a test that covers some functionality. It's worth mentioning that it can't catch every variation of the struct due to the use of UUID.